File: RHP_DeskTop/RHP/Sante/Module_Sante.vb
''' <summary>
''' Module transverse du domaine Sante (RHP_DeskTop).
''' Journal d'acces medical (append-only), controle des fonctions de securite
''' SANTE_CLINIQUE / SANTE_ADMIN / SANTE_AUDIT, verrou CNDP, execution SQL
''' parametree (ADODB.Command) pour les ecritures sur les tables cliniques.
''' </summary>
Module Module_Sante

    ''' <summary>Journal d'acces medical. Ne bloque jamais l'application.</summary>
    Sub Sante_Audit(action As String, objet As String, valeurIndex As String,
                    Optional matriculeConcerne As String = "",
                    Optional succes As Boolean = True,
                    Optional motif As String = "")
        Try
            Dim cmd As New ADODB.Command
            With cmd
                .ActiveConnection = cn
                .CommandType = ADODB.CommandTypeEnum.adCmdText
                .CommandText = "insert into RH_Sante_Audit_Acces " &
                    "(id_Societe, Login_User, id_User, Cod_Profile, Typ_Role, Action, Objet, Valeur_Index, Matricule_Concerne, Poste, IP, Succes, Motif) " &
                    "values (?,?,?,?,?,?,?,?,?,?,?,?,?)"
                .Parameters.Append(.CreateParameter("p1", ADODB.DataTypeEnum.adInteger, ADODB.ParameterDirectionEnum.adParamInput, , Societe.id_Societe))
                .Parameters.Append(.CreateParameter("p2", ADODB.DataTypeEnum.adVarWChar, ADODB.ParameterDirectionEnum.adParamInput, 50, IsNull(theUser.Login, "")))
                .Parameters.Append(.CreateParameter("p3", ADODB.DataTypeEnum.adInteger, ADODB.ParameterDirectionEnum.adParamInput, , IsNull(theUser.id_User, -1)))
                .Parameters.Append(.CreateParameter("p4", ADODB.DataTypeEnum.adVarWChar, ADODB.ParameterDirectionEnum.adParamInput, 10, CStr(IsNull(theUser.Cod_Profile, -1))))
                .Parameters.Append(.CreateParameter("p5", ADODB.DataTypeEnum.adVarWChar, ADODB.ParameterDirectionEnum.adParamInput, 10, IsNull(theUser.Typ_Role, "")))
                .Parameters.Append(.CreateParameter("p6", ADODB.DataTypeEnum.adVarWChar, ADODB.ParameterDirectionEnum.adParamInput, 10, Gauche(action, 10)))
                .Parameters.Append(.CreateParameter("p7", ADODB.DataTypeEnum.adVarWChar, ADODB.ParameterDirectionEnum.adParamInput, 50, Gauche(objet, 50)))
                .Parameters.Append(.CreateParameter("p8", ADODB.DataTypeEnum.adVarWChar, ADODB.ParameterDirectionEnum.adParamInput, 100, Gauche(valeurIndex, 100)))
                .Parameters.Append(.CreateParameter("p9", ADODB.DataTypeEnum.adVarWChar, ADODB.ParameterDirectionEnum.adParamInput, 20, Gauche(matriculeConcerne, 20)))
                .Parameters.Append(.CreateParameter("p10", ADODB.DataTypeEnum.adVarWChar, ADODB.ParameterDirectionEnum.adParamInput, 100, Gauche(IsNull(Environment.MachineName, ""), 100)))
                .Parameters.Append(.CreateParameter("p11", ADODB.DataTypeEnum.adVarWChar, ADODB.ParameterDirectionEnum.adParamInput, 50, ""))
                .Parameters.Append(.CreateParameter("p12", ADODB.DataTypeEnum.adVarWChar, ADODB.ParameterDirectionEnum.adParamInput, 1, IIf(succes, "1", "0")))
                .Parameters.Append(.CreateParameter("p13", ADODB.DataTypeEnum.adVarWChar, ADODB.ParameterDirectionEnum.adParamInput, 250, Gauche(motif, 250)))
                .Execute()
            End With
        Catch ex As Exception
            ' Le journal ne doit jamais interrompre le traitement metier
        End Try
    End Sub

    ''' <summary>Vrai si le profil courant detient la fonction de securite (SANTE_CLINIQUE/ADMIN/AUDIT).
    ''' Le controle s'applique y compris au profil 1 (pas de bypass pour les donnees de sante).</summary>
    Function Sante_FonctionActive(fonction As String) As Boolean
        Try
            If IsNull(theUser.Cod_Profile, 0) = 0 Then Return False
            Dim rs As ADODB.Recordset = CnExecuting(
                "select count(*) from Controle_Droit_Functions where Cod_Profile='" & theUser.Cod_Profile &
                "' and Function_Sec='" & fonction & "' and isnull(Actif,'false')='true'")
            Return CInt(rs.Fields(0).Value) > 0
        Catch ex As Exception
            Return False
        End Try
    End Function

    ''' <summary>Controle d'acces a un domaine sante ; journalise le refus.</summary>
    Function Sante_CheckAccess(domaine As String, objet As String) As Boolean
        Dim fonctions As String()
        Select Case domaine
            Case "CLINIQUE" : fonctions = {"SANTE_CLINIQUE"}
            Case "ADMIN" : fonctions = {"SANTE_ADMIN", "SANTE_CLINIQUE"}
            Case "AUDIT" : fonctions = {"SANTE_AUDIT"}
            Case Else : fonctions = {"SANTE_ADMIN"}
        End Select
        For Each f In fonctions
            If Sante_FonctionActive(f) Then Return True
        Next
        Sante_Audit("AUTH_KO", objet, "", "", False, "Fonction " & domaine & " non accordee")
        Return False
    End Function

    ''' <summary>Verrou de mise en production CNDP : vrai si le traitement doit etre bloque.</summary>
    Function Sante_VerrouCndp() As Boolean
        Try
            Dim rs As ADODB.Recordset = CnExecuting(
                "select dbo.Sys_Sante_Param('BLOCAGE_PROD_SANS_CNDP', " & Societe.id_Societe & ") as b, " &
                "isnull(dbo.Sys_Sante_Param('CNDP_NUM_AUTORISATION', " & Societe.id_Societe & "),'') as a")
            Return (IsNull(rs.Fields("b").Value, "") = "O" AndAlso Trim(IsNull(rs.Fields("a").Value, "")) = "")
        Catch ex As Exception
            Return False
        End Try
    End Function

    ''' <summary>Execution SQL parametree (anti-injection) pour les tables du module.
    ''' params : tableau de tableaux {nom, type ADODB.DataTypeEnum, taille (0 = non fournie, -1 = MAX), valeur}.</summary>
    Function Sante_Execute(sql As String, params As Object(,)) As Boolean
        Try
            Dim cmd As New ADODB.Command
            With cmd
                .ActiveConnection = cn
                .CommandType = ADODB.CommandTypeEnum.adCmdText
                .CommandText = sql
                For i = 0 To params.GetLength(0) - 1
                    Dim taille As Integer = CInt(params(i, 2))
                    Dim valeur As Object = params(i, 3)
                    If valeur Is DBNull.Value Then valeur = Nothing
                    If taille <> 0 Then
                        .Parameters.Append(.CreateParameter(CStr(params(i, 0)), CType(params(i, 1), ADODB.DataTypeEnum), ADODB.ParameterDirectionEnum.adParamInput, taille, valeur))
                    Else
                        .Parameters.Append(.CreateParameter(CStr(params(i, 0)), CType(params(i, 1), ADODB.DataTypeEnum), ADODB.ParameterDirectionEnum.adParamInput, , valeur))
                    End If
                Next
                .Execute()
            End With
            Return True
        Catch ex As Exception
            ErrorMsg(ex)
            Return False
        End Try
    End Function

    ''' <summary>Scalar SQL parametre (une valeur).</summary>
    Function Sante_Scalar(sql As String, params As Object(,)) As Object
        Dim cmd As New ADODB.Command
        With cmd
            .ActiveConnection = cn
            .CommandType = ADODB.CommandTypeEnum.adCmdText
            .CommandText = sql
            For i = 0 To params.GetLength(0) - 1
                Dim taille As Integer = CInt(params(i, 2))
                Dim valeur As Object = params(i, 3)
                If valeur Is DBNull.Value Then valeur = Nothing
                If taille <> 0 Then
                    .Parameters.Append(.CreateParameter(CStr(params(i, 0)), CType(params(i, 1), ADODB.DataTypeEnum), ADODB.ParameterDirectionEnum.adParamInput, taille, valeur))
                Else
                    .Parameters.Append(.CreateParameter(CStr(params(i, 0)), CType(params(i, 1), ADODB.DataTypeEnum), ADODB.ParameterDirectionEnum.adParamInput, , valeur))
                End If
            Next
            Dim rs As ADODB.Recordset = .Execute()
            If rs.EOF Then Return Nothing
            Return rs.Fields(0).Value
        End With
    End Function

    ''' <summary>Numerotation des documents sante : PREFIXE+idSoc+"-"+annee+seq(6).</summary>
    Function Sante_NouveauNumero(prefixe As String, tableName As String, colNum As String, colDate As String) As String
        Dim sql As String = "select isnull(max(racine),0) as racine from (" &
            "select convert(int,case when isnumeric(ISNULL(racine,''))!=1 then 0 else racine end) as racine from " & tableName &
            " outer apply(select RIGHT(" & colNum & ",6) as racine)n " &
            "where id_Societe=" & Societe.id_Societe & " and year(" & colDate & ")=year(getdate()))f"
        Dim rs As ADODB.Recordset = CnExecuting(sql)
        Return prefixe & Societe.id_Societe & "-" & Now.Year & Droite("000000" & CInt(rs.Fields(0).Value + 1), 6)
    End Function

    ''' <summary>Calcul de la prochaine echeance via la fonction du socle.</summary>
    Function Sante_CalculEcheance(matricule As String, datVisite As Date, ByRef codRegle As String) As Object
        codRegle = ""
        Dim rs As ADODB.Recordset = CnExecuting(
            "select Dat_Prochaine_Visite, Cod_Regle_Appliquee from dbo.Sys_Sante_Prochaine_Visite('" &
            matricule & "', " & Societe.id_Societe & ", '" & datVisite.ToString("yyyy-MM-dd") & "')")
        If rs.EOF Then Return Nothing
        codRegle = IsNull(rs.Fields("Cod_Regle_Appliquee").Value, "")
        Return rs.Fields("Dat_Prochaine_Visite").Value
    End Function

    ''' <summary>Libelle d'un parametre reglementaire (avec sa source) pour affichage.</summary>
    Function Sante_Param(codParam As String, Optional defaut As String = "") As String
        Return IsNull(Sante_Scalar("select dbo.Sys_Sante_Param('" & codParam & "', " & Societe.id_Societe & ")", New Object(-1, 3) {}), defaut).ToString()
    End Function
End Module

File: RHP_Portail/rhpBE/modules/module_sante.ts
/* Module transverse du domaine Sante (cloisonnement des donnees medicales).
   Homogene au socle : controle dans les controleurs, base sur les tables
   d'habilitation existantes (Controle_Menu_Functions / Controle_Droit_Functions). */
import { Request, Response } from "express";
import { lireSql } from "./module_sqlRW";
import { Int, NVarChar, DateTime } from "mssql";

export type DomaineSante = "CLINIQUE" | "ADMIN" | "AUDIT";

/* Recupere l'agent connecte (injecte par le middleware validate) */
export function getAgent(req: Request) {
  const { processId, ...theAgent } = req.params as any;
  const idSocNum = Number(theAgent?.id_Societe || 0);
  return { theAgent, idSocNum };
}

/* Id societe valide ou reponse d'erreur (pattern des controleurs existants) */
export function idSocInvalide(res: Response) {
  return res.send({ result: false, message: "id_Societe invalide" });
}

/* Headers interdisant la mise en cache des donnees de sante */
export function setNoStore(res: Response) {
  res.setHeader("Cache-Control", "no-store");
  res.setHeader("Pragma", "no-cache");
}

/* Journal d'acces medical (append-only). Ne bloque jamais la reponse. */
export async function santeAudit(args: {
  req: Request;
  action: "LECT" | "CREA" | "MODI" | "SUPP" | "IMPR" | "EXPO" | "TELE" | "AUTH_KO";
  objet: string;
  valeurIndex?: string;
  matriculeConcerne?: string;
  succes: boolean;
  motif?: string;
}) {
  try {
    const { theAgent, idSocNum } = getAgent(args.req);
    const ip =
      (args.req.headers["x-forwarded-for"] as string)?.split(",")[0]?.trim() ||
      args.req.socket?.remoteAddress ||
      "";
    await lireSql(
      `insert into RH_Sante_Audit_Acces
        (id_Societe, Login_User, id_User, Cod_Profile, Typ_Role, Action, Objet, Valeur_Index,
         Matricule_Concerne, Poste, IP, Succes, Motif)
       values (@p_idSoc, @p_Login, @p_idUser, @p_CodProfile, @p_TypRole, @p_Action, @p_Objet, @p_ValeurIndex,
         @p_Matricule, 'web', @p_IP, @p_Succes, @p_Motif)`,
      [
        { param: "p_idSoc", sqlType: Int, valeur: idSocNum > 0 ? idSocNum : null },
        { param: "p_Login", sqlType: NVarChar, valeur: theAgent?.Login || "" },
        { param: "p_idUser", sqlType: Int, valeur: Number(theAgent?.id_User || -1) },
        { param: "p_CodProfile", sqlType: NVarChar, valeur: String(theAgent?.codProfile ?? "") },
        { param: "p_TypRole", sqlType: NVarChar, valeur: theAgent?.Typ_Role || "" },
        { param: "p_Action", sqlType: NVarChar, valeur: args.action },
        { param: "p_Objet", sqlType: NVarChar, valeur: args.objet },
        { param: "p_ValeurIndex", sqlType: NVarChar, valeur: args.valeurIndex || "" },
        { param: "p_Matricule", sqlType: NVarChar, valeur: args.matriculeConcerne || "" },
        { param: "p_IP", sqlType: NVarChar, valeur: ip.substring(0, 50) },
        { param: "p_Succes", sqlType: NVarChar, valeur: args.succes ? "1" : "0" },
        { param: "p_Motif", sqlType: NVarChar, valeur: (args.motif || "").substring(0, 250) },
      ]
    );
  } catch (e) {
    console.error("[Sante] audit error:", e);
  }
}

/* Controle d'habilitation par domaine, homogene au cloisonnement TeamLeader :
   - CLINIQUE : fonction SANTE_CLINIQUE active pour le profil de l'utilisateur
   - ADMIN    : SANTE_ADMIN ou SANTE_CLINIQUE
   - AUDIT    : SANTE_AUDIT
   Le codProfile vient du JWT (profil Controle_Users rattache par Mail au login). */
export async function checkSanteAccess(
  theAgent: any,
  domaine: DomaineSante
): Promise<{ ok: boolean; motif: string }> {
  const codProfile = String(theAgent?.codProfile ?? "-1");
  if (codProfile === "" || codProfile === "-1" || codProfile === "undefined") {
    return { ok: false, motif: "Profil non rattaché" };
  }
  const fonctions =
    domaine === "CLINIQUE"
      ? "('SANTE_CLINIQUE')"
      : domaine === "ADMIN"
      ? "('SANTE_ADMIN','SANTE_CLINIQUE')"
      : "('SANTE_AUDIT')";
  const rsl = await lireSql(
    `select count(*) as nb from Controle_Droit_Functions
     where Cod_Profile = @profil and Function_Sec in ${fonctions} and isnull(Actif,'false') = 'true'`,
    [{ param: "profil", sqlType: NVarChar, valeur: codProfile }]
  );
  const ok = rsl.result && rsl.data.length > 0 && Number(rsl.data[0].nb) > 0;
  return { ok, motif: ok ? "" : "Fonction " + domaine + " non accordée" };
}

/* Verrou de mise en production CNDP (donnees sensibles de sante) :
   si BLOCAGE_PROD_SANS_CNDP='O' et aucune autorisation renseignee, les ecritures
   cliniques sont bloquees. */
export async function verrouCndpActif(idSocNum: number): Promise<boolean> {
  const rsl = await lireSql(
    `select dbo.Sys_Sante_Param('BLOCAGE_PROD_SANS_CNDP', @idSoc) as blocage,
            isnull(dbo.Sys_Sante_Param('CNDP_NUM_AUTORISATION', @idSoc),'') as autorisation`,
    [{ param: "idSoc", sqlType: Int, valeur: idSocNum }]
  );
  if (!rsl.result || rsl.data.length === 0) return false;
  return rsl.data[0].blocage === "O" && String(rsl.data[0].autorisation).trim() === "";
}

/* Wrapper standard d'un endpoint sante : agent -> idSoc -> habilitation -> audit.
   fn recoit (req, res, idSocNum) et ecrit la reponse. */
export async function santeEndpoint(
  req: Request,
  res: Response,
  domaine: DomaineSante,
  objet: string,
  fn: (req: Request, res: Response, idSocNum: number) => Promise<any>
) {
  setNoStore(res);
  const { theAgent, idSocNum } = getAgent(req);
  if (isNaN(idSocNum) || idSocNum <= 0) {
    idSocInvalide(res);
    return;
  }
  const acces = await checkSanteAccess(theAgent, domaine);
  if (!acces.ok) {
    await santeAudit({
      req,
      action: "AUTH_KO",
      objet,
      succes: false,
      motif: acces.motif,
    });
    res.send({ result: false, message: "Accès non autorisé" });
    return;
  }
  try {
    await fn(req, res, idSocNum);
  } catch (e) {
    console.error("[Sante] endpoint error:", e);
    res.send({ result: false, message: "Erreur interne du serveur" });
  }
}

/* Generation d'un numero de document : <PREFIXE><idSoc>-<annee><seq 6> (pattern socle) */
export async function nouveauNumero(
  prefixe: string,
  tableName: string,
  colNum: string,
  colDate: string,
  idSocNum: number
): Promise<string> {
  const rsl = await lireSql(
    `select '${prefixe}'+convert(nvarchar(10),@idSoc)+'-'+convert(nvarchar(4),year(getdate()))
       +right('000000'+convert(nvarchar(6),isnull(max(racine),0)+1),6) as num
     from (select convert(int,case when isnumeric(ISNULL(racine,''))!=1 then 0 else racine end) as racine
           from ${tableName} outer apply(select RIGHT(${colNum},6) as racine)n
           where id_Societe=@idSoc and year(${colDate})=year(getdate()))f`,
    [{ param: "idSoc", sqlType: Int, valeur: idSocNum }]
  );
  return rsl?.data?.[0]?.num || "";
}

/* Conversion date entree (string ISO ou dd/MM/yyyy) -> objet Date (SmallDateTime) */
export function toDate(v: any): Date | null {
  if (!v) return null;
  if (v instanceof Date) return isNaN(v.getTime()) ? null : v;
  const s = String(v);
  if (/^\d{2}\/\d{2}\/\d{4}/.test(s)) {
    const [d, m, y] = s.split("/");
    const dt = new Date(Number(y), Number(m) - 1, Number(d));
    return isNaN(dt.getTime()) ? null : dt;
  }
  const dt = new Date(s);
  return isNaN(dt.getTime()) ? null : dt;
}
